### ISSUE TITLE
Got MemoryError when copy huge local file to remote

### ISSUE THREAD

**Issue by nyasukun** [ansible/ansible#16391]
##### ISSUE TYPE

- Bug Report
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### OS / ENVIRONMENT

Ubuntu 14.04.4 LTS
##### SUMMARY

Got MemoryError when copy huge local file to remote.
##### STEPS TO REPRODUCE

```
$ cat huge_copy.yml 

---
- hosts: all
  become: yes

  tasks:
    - name: Copy huge file
      copy: src=5gb dest=dst

$ dd if=/dev/zero of=5gb bs=1G count=5
5+0 records in
5+0 records out
5368709120 bytes (5.4 GB) copied, 206.341 s, 26.0 MB/s

$ ansible-playbook -i hosts huge_copy.yml
```

##### EXPECTED RESULTS

Copied successfully.
##### ACTUAL RESULTS

```
$ ansible-playbook -i hosts huge_copy.yml -vvv
Using /etc/ansible/ansible.cfg as config file

PLAYBOOK: huge_copy.yml ********************************************************
1 plays in huge_copy.yml

PLAY [all] *********************************************************************

TASK [setup] *******************************************************************
<192.168.80.129> ESTABLISH SSH CONNECTION FOR USER: ansible
<192.168.80.129> SSH: EXEC ssh -C -q -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ansible -o ConnectTimeout=10 -o ControlPath=/home/y/.ansible/cp/ansible-ssh-%h-%p-%r 192.168.80.129 '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1466567672.43-71840509019686 `" && echo ansible-tmp-1466567672.43-71840509019686="` echo $HOME/.ansible/tmp/ansible-tmp-1466567672.43-71840509019686 `" ) && sleep 0'"'"''
<192.168.80.129> PUT /tmp/tmp3Bk56k TO /home/ansible/.ansible/tmp/ansible-tmp-1466567672.43-71840509019686/setup
<192.168.80.129> SSH: EXEC sftp -b - -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ansible -o ConnectTimeout=10 -o ControlPath=/home/y/.ansible/cp/ansible-ssh-%h-%p-%r '[192.168.80.129]'
<192.168.80.129> ESTABLISH SSH CONNECTION FOR USER: ansible
<192.168.80.129> SSH: EXEC ssh -C -q -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ansible -o ConnectTimeout=10 -o ControlPath=/home/y/.ansible/cp/ansible-ssh-%h-%p-%r -tt 192.168.80.129 '/bin/sh -c '"'"'sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-rmwgnaejatjdbjwnltokljxsdjjlhagx; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/ansible/.ansible/tmp/ansible-tmp-1466567672.43-71840509019686/setup; rm -rf "/home/ansible/.ansible/tmp/ansible-tmp-1466567672.43-71840509019686/" > /dev/null 2>&1'"'"'"'"'"'"'"'"' && sleep 0'"'"''
ok: [192.168.80.129]

TASK [Copy huge file] **********************************************************
task path: /home/y/nfvdv4-single/huge_copy.yml:6
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/ansible/executor/task_executor.py", line 124, in run
    res = self._execute()
  File "/usr/lib/python2.7/dist-packages/ansible/executor/task_executor.py", line 446, in _execute
    result = self._handler.run(task_vars=variables)
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/action/copy.py", line 156, in run
    source_full = self._loader.get_real_file(source_full)
  File "/usr/lib/python2.7/dist-packages/ansible/parsing/dataloader.py", line 334, in get_real_file
    data = f.read()
MemoryError

fatal: [192.168.80.129]: FAILED! => {"failed": true, "msg": "Unexpected failure during module execution.", "stdout": ""}

NO MORE HOSTS LEFT *************************************************************
    to retry, use: --limit @huge_copy.retry

PLAY RECAP *********************************************************************
192.168.80.129             : ok=1    changed=0    unreachable=0    failed=1  
```

**Comment by ferrants** [comment]
Seeing this in 2.1 as well, seems like a regression from what was working (for me) in 1.9.2.

**Comment by ferrants** [comment]
I've verified that https://github.com/ansible/ansible/pull/16392 is a fix for this issue. Waiting on it to be merged.

**Comment by gnosek** [comment]
Has this been released/backported to 2.1.x? I'm running fresh 2.1.2.0-1ppa~trusty from the official PPA and still get MemoryError (understandable as the code is not fixed there).

In this particular install, I'd _really_ like to stick with the PPA.

**Comment by grizmin** [comment]
It has not been fixed anywhere.

root@ansible:/etc/ansible/playbooks/javanodes[env: qa1]# ansible --version
ansible 2.2.0 (devel 15ee97d214) last updated 2016/10/03 13:05:00 (GMT +000)
  lib/ansible/modules/core: (detached HEAD 7de287237f) last updated 2016/07/17 11:35:06 (GMT +000)
  lib/ansible/modules/extras: (detached HEAD 68ca157f3b) last updated 2016/07/17 11:35:10 (GMT +000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

root@ansible:/etc/ansible/playbooks/javanodes[env: qa1]# ansible qansible qa1-feeder3 -m copy -a "src=/root/.ansible/assets/jdk-8u66-linux-x64.tar.gz dest=/tmp/jdk-8u66-linux-x64.tar.gz" -vvv
Using /etc/ansible/ansible.cfg as config file
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
File "/root/ansible/lib/ansible/executor/task_executor.py", line 122, in run
res = self.execute()
File "/root/ansible/lib/ansible/executor/task_executor.py", line 488, in execute
result = self._handler.run(task_vars=variables)
File "/root/ansible/lib/ansible/plugins/action/copy.py", line 153, in run
source_full = self._loader.get_real_file(source_full)
File "/root/ansible/lib/ansible/parsing/dataloader.py", line 398, in get_real_file
if is_encrypted_file(f):
File "/root/ansible/lib/ansible/parsing/vault/__init.py", line 152, in is_encrypted_file
b_vaulttext = to_bytes(to_text(vaulttext, encoding='ascii', errors='strict'), encoding='ascii', errors='strict')
File "/root/ansible/lib/ansible/module_utils/_text.py", line 183, in to_text
return obj.decode(encoding, errors)
MemoryError

qa1-feeder3 | FAILED! => {
"failed": true,
"msg": "Unexpected failure during module execution.",
"stdout": ""
}
root@ansible:/etc/ansible/playbooks/javanodes[env: qa1]#

-rw-r--r-- 1 root root 181287376 Jul 19 21:45 jdk-8u66-linux-x64.tar.gz (173M)

VM is with 1G ram running debian used only for ansible. 
root@ansible:/etc/ansible/playbooks/javanodes[env: qa1]# cat /proc/sys/vm/overcommit_memory
0
root@ansible:/etc/ansible/playbooks/javanodes[env: qa1]#

**Comment by ak5** [comment]
I am running into this for the first time _after_ upgrading to 2.2 via ubuntu ppa :(

I am trying to copy a saved docker image compressed with bz2 to a remote host in a role included in a playbook.

```
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/ansible/executor/task_executor.py", line 119, in run
    res = self._execute()
  File "/usr/lib/python2.7/dist-packages/ansible/executor/task_executor.py", line 490, in _execute
    result = self._handler.run(task_vars=variables)
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/action/copy.py", line 153, in run
    source_full = self._loader.get_real_file(source_full)
  File "/usr/lib/python2.7/dist-packages/ansible/parsing/dataloader.py", line 402, in get_real_file
    if is_encrypted_file(f):
  File "/usr/lib/python2.7/dist-packages/ansible/parsing/vault/__init__.py", line 150, in is_encrypted_file
    vaulttext = file_obj.read(count)
MemoryError

fatal: [XXX.XXX.XXX.XXX]: FAILED! => {
    "failed": true, 
    "msg": "Unexpected failure during module execution.", 
    "stdout": ""
}

```

**Comment by ferrants** [comment]
See if this change fixes it for you guys: https://github.com/ansible/ansible/pull/18365

**Comment by abadger** [comment]
This should be fixed in the devel and stable-2.2 branches via #18368

In the stable-2.1 branch found out that the original fix wasn't backported.  Cherry-picked adea1f2b80d806d94ca1bdb2d06f2df077feb948 to fix it there.

That means that the fix will show up in 2.3.0, 2.2.1, and, if we make another 2.1 release, 2.1.4.

Chronologically, 2.2.1 is most likely the next release that we'll make.

And just a side note: It's better to open a new bug report than to comment on a closed one.  Our workflow around github issues makes it easy to miss new comments on closed issues.  Thanks!

**Comment by rishibamba** [comment]
Same issue on Ansible 2.2.0.0. Directory size around 3.2GB
